File: src/lib/workspace/tab.ts
export type TabKind =
  | 'welcome'
  | 'new-tab'
  | 'timeline'
  | 'global'
  | 'public-chat'
  | 'notifications'
  | 'profile'
  | 'profile-edit'
  | 'upload-settings'
  | 'account-manager'
  | 'npub-miner'
  | 'thread'
  | 'relay-monitor'
  | 'relay-settings'
  | 'network-stats'
  | 'search'
  | 'custom-request'
  | 'author-context'
  | 'tweet'
  | 'settings';

export type WorkspaceTab = {
  readonly id: string;
  readonly kind: TabKind;
  readonly title: string;
  readonly icon: string;
  readonly avatarUrl?: string | null;
  readonly dirty?: boolean;
  readonly loading?: boolean;
  readonly accountId?: string | null;
  readonly config: Record<string, unknown>;
  readonly state: Record<string, unknown>;
  readonly createdAt: number;
  readonly updatedAt: number;
};

export function createTab(
  kind: TabKind,
  title: string,
  config: Record<string, unknown> = {},
): WorkspaceTab {
  const now = Date.now();
  return {
    id: crypto.randomUUID(),
    kind,
    title,
    icon: iconFor(kind),
    config,
    state: {},
    createdAt: now,
    updatedAt: now,
  };
}

export function iconFor(kind: TabKind): string {
  const icons: Record<TabKind, string> = {
    welcome: 'star',
    'new-tab': 'plus',
    timeline: 'list',
    global: 'globe',
    'public-chat': 'message',
    notifications: 'bell',
    profile: 'user',
    'profile-edit': 'id-card',
    'upload-settings': 'upload',
    'account-manager': 'users',
    'npub-miner': 'pickaxe',
    thread: 'message',
    'relay-monitor': 'radio',
    'relay-settings': 'sliders',
    'network-stats': 'activity',
    search: 'search',
    'custom-request': 'braces',
    'author-context': 'user-round-search',
    tweet: 'edit',
    settings: 'settings',
  };
  return icons[kind];
}

export function touchTab(tab: WorkspaceTab): WorkspaceTab {
  return { ...tab, updatedAt: Date.now() };
}

export function convertTab(
  tab: WorkspaceTab,
  kind: TabKind,
  title: string,
  config: Record<string, unknown> = {},
): WorkspaceTab {
  return {
    ...tab,
    kind,
    title,
    icon: iconFor(kind),
    config,
    state: {},
    updatedAt: Date.now(),
  };
}

File: crates/lkjstr-domain/src/workspace/tab.rs
#![doc = "Workspace tabs."]

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum TabKind {
    Welcome,
    NewTab,
    Timeline,
    Global,
    PublicChat,
    Notifications,
    Profile,
    ProfileEdit,
    UploadSettings,
    AccountManager,
    NpubMiner,
    Thread,
    RelayMonitor,
    RelaySettings,
    NetworkStats,
    Search,
    CustomRequest,
    AuthorContext,
    Tweet,
    Settings,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceTab {
    pub id: String,
    pub kind: TabKind,
    pub title: String,
    pub icon: String,
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub config: BTreeMap<String, String>,
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub state: BTreeMap<String, String>,
    pub created_at: u64,
    pub updated_at: u64,
}

impl WorkspaceTab {
    #[must_use]
    pub fn new(id: impl Into<String>, kind: TabKind, now: u64) -> Self {
        let title = title_for(kind).to_owned();
        let icon = icon_for(kind).to_owned();
        Self {
            id: id.into(),
            kind,
            title,
            icon,
            config: BTreeMap::new(),
            state: BTreeMap::new(),
            created_at: now,
            updated_at: now,
        }
    }

    #[must_use]
    pub fn with_config(mut self, config: BTreeMap<String, String>, now: u64) -> Self {
        self.config = config;
        self.updated_at = now;
        self
    }
}

#[must_use]
pub const fn title_for(kind: TabKind) -> &'static str {
    match kind {
        TabKind::Welcome => "Welcome",
        TabKind::NewTab => "New Tab",
        TabKind::Timeline => "Home",
        TabKind::Global => "Global",
        TabKind::PublicChat => "Public Chat",
        TabKind::Notifications => "Notifications",
        TabKind::Profile => "Profile",
        TabKind::ProfileEdit => "Profile Edit",
        TabKind::UploadSettings => "Upload Settings",
        TabKind::AccountManager => "Accounts",
        TabKind::NpubMiner => "Mine npub",
        TabKind::Thread => "Thread",
        TabKind::RelayMonitor => "lkjstr Log",
        TabKind::RelaySettings => "Relay Settings",
        TabKind::NetworkStats => "Stats",
        TabKind::Search => "Search",
        TabKind::CustomRequest => "Custom Request",
        TabKind::AuthorContext => "Author Context",
        TabKind::Tweet => "Tweet",
        TabKind::Settings => "Settings",
    }
}

#[must_use]
pub const fn tab_kind_key(kind: TabKind) -> &'static str {
    match kind {
        TabKind::Welcome => "welcome",
        TabKind::NewTab => "new-tab",
        TabKind::Timeline => "timeline",
        TabKind::Global => "global",
        TabKind::PublicChat => "public-chat",
        TabKind::Notifications => "notifications",
        TabKind::Profile => "profile",
        TabKind::ProfileEdit => "profile-edit",
        TabKind::UploadSettings => "upload-settings",
        TabKind::AccountManager => "account-manager",
        TabKind::NpubMiner => "npub-miner",
        TabKind::Thread => "thread",
        TabKind::RelayMonitor => "relay-monitor",
        TabKind::RelaySettings => "relay-settings",
        TabKind::NetworkStats => "network-stats",
        TabKind::Search => "search",
        TabKind::CustomRequest => "custom-request",
        TabKind::AuthorContext => "author-context",
        TabKind::Tweet => "tweet",
        TabKind::Settings => "settings",
    }
}

#[must_use]
pub const fn icon_for(kind: TabKind) -> &'static str {
    match kind {
        TabKind::Welcome => "star",
        TabKind::NewTab => "plus",
        TabKind::Timeline => "list",
        TabKind::Global => "globe",
        TabKind::PublicChat => "message",
        TabKind::Notifications => "bell",
        TabKind::Profile => "user",
        TabKind::ProfileEdit => "id-card",
        TabKind::UploadSettings => "upload",
        TabKind::AccountManager => "users",
        TabKind::NpubMiner => "pickaxe",
        TabKind::Thread => "message",
        TabKind::RelayMonitor => "radio",
        TabKind::RelaySettings => "sliders",
        TabKind::NetworkStats => "activity",
        TabKind::Search => "search",
        TabKind::CustomRequest => "braces",
        TabKind::AuthorContext => "user-round-search",
        TabKind::Tweet => "edit",
        TabKind::Settings => "settings",
    }
}
